python/semantic_kernel/agents/agent.py

```diff
@@ -478,3 +560,336 @@ def as_mcp_server(
             instructions=instructions,
             lifespan=lifespan,
         )
+
+
+# region Declarative Spec Handling
+
+
+@runtime_checkable
+class DeclarativeSpecProtocol(Protocol):
+    """Protocol for declarative spec mixin."""
+
+    @classmethod
+    def resolve_placeholders(
+        cls: type,
+        yaml_str: str,
+        settings: "KernelBaseSettings | None" = None,
+        extras: dict[str, Any] | None = None,
+    ) -> str:
+        """Resolve placeholders in the YAML string."""
+        ...
+
+    @classmethod
+    async def from_yaml(
+        cls: type,
+        yaml_str: str,
+        *,
+        kernel: Kernel,
+        settings: "KernelBaseSettings | None" = None,
+        extras: dict[str, Any] | None = None,
+        **kwargs,
+    ) -> "Agent":
+        """Create an agent instance from a YAML string."""
+        ...
+
+    @classmethod
+    async def from_dict(
+        cls: type,
+        data: dict,
+        *,
+        kernel: Kernel,
+        settings: "KernelBaseSettings | None" = None,
+        **kwargs,
+    ) -> "Agent":
+        """Create an agent from a dictionary."""
+        ...
+
+
+_TAgent = TypeVar("_TAgent", bound=Agent)
+
+# Global agent type registry
+AGENT_TYPE_REGISTRY: dict[str, type[Agent]] = {}
+
+
+def register_agent_type(agent_type: str):
+    """Decorator to register an agent type with the registry.
+
+    Example usage:
+        @register_agent_type("my_custom_agent")
+        class MyCustomAgent(Agent):
+            ...
+    """
+
+    def decorator(cls: type[_TAgent]) -> type[_TAgent]:
+        AGENT_TYPE_REGISTRY[agent_type.lower()] = cls
+        return cls
+
+    return decorator
+
+
+class AgentRegistry:
+    """Responsible for creating agents from YAML, dicts, or files."""
+
+    @staticmethod
+    def register_type(agent_type: str, agent_cls: type[Agent]) -> None:
+        """Register a new agent type at runtime.
+
+        Args:
+            agent_type: The string identifier representing the agent type (e.g., 'chat_completion_agent').
+            agent_cls: The class implementing the agent, inheriting from `Agent`.
+
+        Example:
+            AgentRegistry.register_type("my_custom_agent", MyCustomAgent)
+        """
+        AGENT_TYPE_REGISTRY[agent_type.lower()] = agent_cls
+
+    @staticmethod
+    async def create_from_yaml(
+        yaml_str: str,
```

Comment:
this assume we already have a string, should we add a convenience function that reads from file and calls this?